File: digital-marketing-GO/backend/internal/handlers/stats.go
package handlers

import (
	"context"
	"net/http"

	"github.com/jackc/pgx/v5/pgxpool"
)

type StatsHandler struct {
	DB *pgxpool.Pool
}

func (h *StatsHandler) PublicStats(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()

	stats := map[string]interface{}{
		"creators":  countWhere(ctx, h.DB, "users", "role='influencer'"),
		"brands":    countWhere(ctx, h.DB, "users", "role='brand'"),
		"campaigns": countWhere(ctx, h.DB, "products", "1=1"),
		"bids":      countWhere(ctx, h.DB, "bids", "1=1"),
		"approved":  countWhere(ctx, h.DB, "bids", "status='approved'"),
	}

	approved := stats["approved"].(int)
	delivered := 0
	if approved > 0 {
		delivered = countWhere(ctx, h.DB, "concepts", "1=1")
	}
	if approved > 0 {
		stats["delivery"] = (delivered * 100) / approved
	} else {
		stats["delivery"] = 100
	}

	writeJSON(w, http.StatusOK, stats)
}

func countWhere(ctx context.Context, db *pgxpool.Pool, table, where string) int {
	var n int
	_ = db.QueryRow(ctx, "SELECT COUNT(*) FROM "+table+" WHERE "+where).Scan(&n)
	return n
}

File: digital-marketing/backend/src/handlers/stats.rs
use axum::{extract::State, response::IntoResponse, Json};
use serde_json::json;

use crate::{errors::AppResult, AppState};

#[derive(sqlx::FromRow)]
struct CountRow {
    cnt: i64,
}

async fn count(db: &sqlx::PgPool, sql: &str) -> Result<i64, sqlx::Error> {
    let row: CountRow = sqlx::query_as(sql).fetch_one(db).await?;
    Ok(row.cnt)
}

pub async fn public_stats(State(state): State<AppState>) -> AppResult<impl IntoResponse> {
    let creators =
        count(&state.db, "SELECT COUNT(*) AS cnt FROM users WHERE role='influencer'").await?;
    let brands =
        count(&state.db, "SELECT COUNT(*) AS cnt FROM users WHERE role='brand'").await?;
    let campaigns = count(&state.db, "SELECT COUNT(*) AS cnt FROM products").await?;
    let bids = count(&state.db, "SELECT COUNT(*) AS cnt FROM bids").await?;
    let approved =
        count(&state.db, "SELECT COUNT(*) AS cnt FROM bids WHERE status='approved'").await?;
    let delivered = count(&state.db, "SELECT COUNT(*) AS cnt FROM concepts").await?;

    let delivery = if approved > 0 {
        ((delivered as f64 / approved as f64) * 100.0).round() as i64
    } else {
        100
    };

    Ok(Json(json!({
        "creators": creators,
        "brands": brands,
        "campaigns": campaigns,
        "bids": bids,
        "approved": approved,
        "delivery": delivery,
    })))
}
